File: tests/matrix/javascript/test_matrix.mjs
import assert from "node:assert/strict";
import fs from "node:fs";
import { unique, sum, average, groupBy } from "../../../utilities/collections/javascript/collection_utils.js";
import { isEmpty, capitalize, reverseString, isValidEmail, normalizeWhitespace, toTitleCase } from "../../../utilities/strings/javascript/string_utils.js";
import { clamp, isNumeric, minValue, maxValue } from "../../../utilities/numbers/javascript/basic_utils.js";
import { parseDate, daysBetween } from "../../../utilities/dates/javascript/basic_utils.js";
import { isNonEmpty, isEmail, isWithinRange } from "../../../utilities/validation/javascript/basic_utils.js";

const cases = JSON.parse(fs.readFileSync(new URL("../fixtures/core-utility-cases.json", import.meta.url), "utf8"));

assert.deepEqual(unique(cases.collections.unique.input), cases.collections.unique.expected);
assert.equal(sum(cases.collections.sum.input), cases.collections.sum.expected);
assert.equal(average(cases.collections.average.input), cases.collections.average.expected);
assert.equal(average(cases.collections.emptyAverage.input), cases.collections.emptyAverage.expected);
assert.deepEqual(groupBy([{ group: "a", value: 1 }, { group: "b", value: 2 }, { group: "a", value: 3 }], "group"), {
  a: [{ group: "a", value: 1 }, { group: "a", value: 3 }],
  b: [{ group: "b", value: 2 }]
});

assert.equal(isEmpty(""), true);
assert.equal(capitalize(cases.strings.capitalize.input), cases.strings.capitalize.expected);
assert.equal(reverseString(cases.strings.reverse.input), cases.strings.reverse.expected);
assert.equal(isValidEmail(cases.strings.validEmail.input), cases.strings.validEmail.expected);
assert.equal(isValidEmail(cases.strings.invalidEmail.input), cases.strings.invalidEmail.expected);
assert.equal(normalizeWhitespace(cases.strings.normalizeWhitespace.input), cases.strings.normalizeWhitespace.expected);
assert.equal(toTitleCase(cases.strings.titleCase.input), cases.strings.titleCase.expected);

assert.equal(clamp(...cases.numbers.clampLow.input), cases.numbers.clampLow.expected);
assert.equal(clamp(...cases.numbers.clampHigh.input), cases.numbers.clampHigh.expected);
assert.equal(clamp(...cases.numbers.clampBoundary.input), cases.numbers.clampBoundary.expected);
assert.equal(isNumeric(cases.numbers.numeric.input), cases.numbers.numeric.expected);
assert.equal(isNumeric(cases.numbers.nonNumeric.input), cases.numbers.nonNumeric.expected);
assert.equal(minValue(cases.numbers.min.input), cases.numbers.min.expected);
assert.equal(maxValue(cases.numbers.max.input), cases.numbers.max.expected);

assert.ok(parseDate(cases.dates.valid.input));
assert.equal(parseDate(cases.dates.invalid.input), undefined);
assert.equal(daysBetween(cases.dates.daysBetween.start, cases.dates.daysBetween.end), cases.dates.daysBetween.expected);
assert.ok(parseDate(cases.dates.localDateTime.input));

assert.equal(isNonEmpty(cases.validation.nonEmpty.input), cases.validation.nonEmpty.expected);
assert.equal(isNonEmpty(cases.validation.empty.input), cases.validation.empty.expected);
assert.equal(isEmail(cases.validation.validEmail.input), cases.validation.validEmail.expected);
assert.equal(isEmail(cases.validation.invalidEmail.input), cases.validation.invalidEmail.expected);
assert.equal(isWithinRange(...cases.validation.rangeLowBoundary.input), cases.validation.rangeLowBoundary.expected);
assert.equal(isWithinRange(...cases.validation.rangeHighBoundary.input), cases.validation.rangeHighBoundary.expected);
assert.equal(isWithinRange(...cases.validation.rangeOutside.input), cases.validation.rangeOutside.expected);

File: tests/matrix/python/test_matrix.py
import importlib.util
import json
import pathlib
import unittest

ROOT = pathlib.Path(__file__).resolve().parents[3]

def load_module(path):
    spec = importlib.util.spec_from_file_location(path.stem, path)
    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)
    return module

cases = json.loads((ROOT / "tests/matrix/fixtures/core-utility-cases.json").read_text())
collections = load_module(ROOT / "utilities/collections/python/collection_utils.py")
strings = load_module(ROOT / "utilities/strings/python/string_utils.py")
numbers = load_module(ROOT / "utilities/numbers/python/basic_utils.py")
dates = load_module(ROOT / "utilities/dates/python/basic_utils.py")
validation = load_module(ROOT / "utilities/validation/python/basic_utils.py")

class CrossLanguageMatrixTest(unittest.TestCase):
    def test_collections(self):
        c = cases["collections"]
        self.assertEqual(collections.unique(c["unique"]["input"]), c["unique"]["expected"])
        self.assertEqual(collections.sum_values(c["sum"]["input"]), c["sum"]["expected"])
        self.assertEqual(collections.average(c["average"]["input"]), c["average"]["expected"])
        self.assertEqual(collections.average(c["emptyAverage"]["input"]), c["emptyAverage"]["expected"])

    def test_strings(self):
        s = cases["strings"]
        self.assertEqual(strings.capitalize(s["capitalize"]["input"]), s["capitalize"]["expected"])
        self.assertEqual(strings.reverse_string(s["reverse"]["input"]), s["reverse"]["expected"])
        self.assertEqual(strings.is_valid_email(s["validEmail"]["input"]), s["validEmail"]["expected"])
        self.assertEqual(strings.is_valid_email(s["invalidEmail"]["input"]), s["invalidEmail"]["expected"])
        self.assertEqual(strings.normalize_whitespace(s["normalizeWhitespace"]["input"]), s["normalizeWhitespace"]["expected"])
        self.assertEqual(strings.to_title_case(s["titleCase"]["input"]), s["titleCase"]["expected"])

    def test_numbers(self):
        n = cases["numbers"]
        self.assertEqual(numbers.clamp(*n["clampLow"]["input"]), n["clampLow"]["expected"])
        self.assertEqual(numbers.clamp(*n["clampHigh"]["input"]), n["clampHigh"]["expected"])
        self.assertEqual(numbers.clamp(*n["clampBoundary"]["input"]), n["clampBoundary"]["expected"])
        self.assertEqual(numbers.is_numeric(n["numeric"]["input"]), n["numeric"]["expected"])
        self.assertEqual(numbers.is_numeric(n["nonNumeric"]["input"]), n["nonNumeric"]["expected"])
        self.assertEqual(numbers.min_value(n["min"]["input"]), n["min"]["expected"])
        self.assertEqual(numbers.max_value(n["max"]["input"]), n["max"]["expected"])

    def test_dates(self):
        d = cases["dates"]
        self.assertIsNotNone(dates.parse_date(d["valid"]["input"]))
        self.assertIsNone(dates.parse_date(d["invalid"]["input"]))
        self.assertEqual(dates.days_between(d["daysBetween"]["start"], d["daysBetween"]["end"]), d["daysBetween"]["expected"])
        self.assertIsNotNone(dates.parse_date(d["localDateTime"]["input"]))

    def test_validation(self):
        v = cases["validation"]
        self.assertEqual(validation.is_non_empty(v["nonEmpty"]["input"]), v["nonEmpty"]["expected"])
        self.assertEqual(validation.is_non_empty(v["empty"]["input"]), v["empty"]["expected"])
        self.assertEqual(validation.is_email(v["validEmail"]["input"]), v["validEmail"]["expected"])
        self.assertEqual(validation.is_email(v["invalidEmail"]["input"]), v["invalidEmail"]["expected"])
        self.assertEqual(validation.is_within_range(*v["rangeLowBoundary"]["input"]), v["rangeLowBoundary"]["expected"])
        self.assertEqual(validation.is_within_range(*v["rangeHighBoundary"]["input"]), v["rangeHighBoundary"]["expected"])
        self.assertEqual(validation.is_within_range(*v["rangeOutside"]["input"]), v["rangeOutside"]["expected"])

if __name__ == "__main__":
    unittest.main()
